# Module 02 Homework - Pandas
Using a dataset from _Wine Spectator_, a wine magazine, we will practice the material that we covered in notebook M02_Colab04 to M02_Colab06. This notebook will cover data transformation, grouping, and sorting using pandas.

Created by Shannon Lo 017287452
Last updated: Sunday, Feb 22, 2026

In [15]:
csvurl="https://gist.githubusercontent.com/clairehq/" + \
        "79acab35be50eaf1c383948ed3fd1129/raw/407a02139ae1e134992b90b4b2b8c329b3d73a6a/winemag-data-130k-v2.csv"

import pandas as pd
wine = pd.read_csv(csvurl)

**Data cleaning**  
Notice that the first column is redundant. Part of data analysis is cleaning and removing redundancy. How would you drop the redundant column inplace, that is overwrite the dataframe.

In [19]:
wine.info() # shows that there is a column named unnamed

# drop the column named "Unnamed: 0" to clean the data
wine.drop(columns=['Unnamed: 0'], inplace=True)

# .head() shows that the redundant column has now been removed
wine.head()

<class 'pandas.DataFrame'>
RangeIndex: 65499 entries, 0 to 65498
Data columns (total 14 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   Unnamed: 0             65499 non-null  int64  
 1   country                65467 non-null  str    
 2   description            65499 non-null  str    
 3   designation            46588 non-null  str    
 4   points                 65499 non-null  int64  
 5   price                  60829 non-null  float64
 6   province               65467 non-null  str    
 7   region_1               54744 non-null  str    
 8   region_2               25170 non-null  str    
 9   taster_name            51856 non-null  str    
 10  taster_twitter_handle  49467 non-null  str    
 11  title                  65499 non-null  str    
 12  variety                65499 non-null  str    
 13  winery                 65499 non-null  str    
dtypes: float64(1), int64(2), str(11)
memory usage: 7.0 MB


country                                        description  \
0     Italy  Aromas include tropical fruit, broom, brimston...   
1  Portugal  This is ripe and fruity, a wine that is smooth...   
2        US  Tart and snappy, the flavors of lime flesh and...   
3        US  Pineapple rind, lemon pith and orange blossom ...   
4        US  Much like the regular bottling from 2012, this...   

                          designation  points  price           province  \
0                        Vulkà Bianco      87    NaN  Sicily & Sardinia   
1                            Avidagos      87   15.0              Douro   
2                                 NaN      87   14.0             Oregon   
3                Reserve Late Harvest      87   13.0           Michigan   
4  Vintner's Reserve Wild Child Block      87   65.0             Oregon   

              region_1           region_2         taster_name  \
0                 Etna                NaN       Kerin O’Keefe   
1                  NaN                NaN          Roger Voss   
2    Willamette Valley  Willamette Valley        Paul Gregutt   
3  Lake Michigan Shore                NaN  Alexander Peartree   
4    Willamette Valley  Willamette Valley        Paul Gregutt   

  taster_twitter_handle                                              title  \
0          @kerinokeefe                  Nicosia 2013 Vulkà Bianco  (Etna)   
1            @vossroger      Quinta dos Avidagos 2011 Avidagos Red (Douro)   
2           @paulgwine       Rainstorm 2013 Pinot Gris (Willamette Valley)   
3                   NaN  St. Julian 2013 Reserve Late Harvest Riesling ...   
4           @paulgwine   Sweet Cheeks 2012 Vintner's Reserve Wild Child...   

          variety               winery  
0     White Blend              Nicosia  
1  Portuguese Red  Quinta dos Avidagos  
2      Pinot Gris            Rainstorm  
3        Riesling           St. Julian  
4      Pinot Noir         Sweet Cheeks

#### Question 1: ####  
What is the mean of the points column?

In [26]:
# Use built-in method of pandas to determine mean
wine["points"].mean()


np.float64(88.43403716087269)

#### Question 2: ####  
How many countries are present in this dataset? (Only count each country once)

In [34]:
# use .nunique() to count unique country entries
wine["country"].nunique()

41

#### Question 3: ####
How many times does each country appeared in this dataset? Show each country and the corresponding count (show counts in ascending order)

In [38]:
# use .value_counts() to count number of entries each country has. Ascending orders it from smallest to largest.
wine["country"].value_counts(ascending=True)

# The US is mentioned the most times within the dataset, followed my France, Italy, and Spain.

country
Armenia                       1
Bosnia and Herzegovina        1
Slovakia                      1
Luxembourg                    4
India                         4
Switzerland                   4
Ukraine                       5
Czech Republic                6
Cyprus                        6
Macedonia                     6
Serbia                        7
Peru                          8
Morocco                      11
Lebanon                      20
Moldova                      30
Mexico                       31
Brazil                       31
England                      36
Georgia                      37
Slovenia                     39
Turkey                       43
Croatia                      44
Hungary                      61
Uruguay                      61
Romania                      67
Bulgaria                     68
Canada                      108
Greece                      242
Israel                      259
New Zealand                 733
South Africa                737


#### Question 4: ####
Create a variable `adjusted_price` containing the adjusted price which is the price subtracted by the average price. *This is called **"centering" transformation** - a method commonly used in the preprocessing step before applying various machine learning algorithms.*

In [41]:
# Price subtracted by average price
adjusted_price = wine['price'] - wine['price'].mean()
print(adjusted_price)

# use .head() again to view
wine.head()

0              NaN
1       -20.232932
2       -21.232932
3       -22.232932
4        29.767068
           ...    
65494     9.767068
65495   -13.232932
65496   -15.232932
65497    -4.232932
65498   -25.232932
Name: price, Length: 65499, dtype: float64


country                                        description  \
0     Italy  Aromas include tropical fruit, broom, brimston...   
1  Portugal  This is ripe and fruity, a wine that is smooth...   
2        US  Tart and snappy, the flavors of lime flesh and...   
3        US  Pineapple rind, lemon pith and orange blossom ...   
4        US  Much like the regular bottling from 2012, this...   

                          designation  points  price           province  \
0                        Vulkà Bianco      87    NaN  Sicily & Sardinia   
1                            Avidagos      87   15.0              Douro   
2                                 NaN      87   14.0             Oregon   
3                Reserve Late Harvest      87   13.0           Michigan   
4  Vintner's Reserve Wild Child Block      87   65.0             Oregon   

              region_1           region_2         taster_name  \
0                 Etna                NaN       Kerin O’Keefe   
1                  NaN                NaN          Roger Voss   
2    Willamette Valley  Willamette Valley        Paul Gregutt   
3  Lake Michigan Shore                NaN  Alexander Peartree   
4    Willamette Valley  Willamette Valley        Paul Gregutt   

  taster_twitter_handle                                              title  \
0          @kerinokeefe                  Nicosia 2013 Vulkà Bianco  (Etna)   
1            @vossroger      Quinta dos Avidagos 2011 Avidagos Red (Douro)   
2           @paulgwine       Rainstorm 2013 Pinot Gris (Willamette Valley)   
3                   NaN  St. Julian 2013 Reserve Late Harvest Riesling ...   
4           @paulgwine   Sweet Cheeks 2012 Vintner's Reserve Wild Child...   

          variety               winery  
0     White Blend              Nicosia  
1  Portuguese Red  Quinta dos Avidagos  
2      Pinot Gris            Rainstorm  
3        Riesling           St. Julian  
4      Pinot Noir         Sweet Cheeks

#### Question 5: ####
What is the title of the wine that has the highest points-to-price ratio in the dataset?

In [46]:
# Title of Wine

# create a new column "ratio" to document points-to-price
wine['ratio'] = wine['points'] / wine['price']

highest_points_to_price = wine.loc[wine['ratio'].idxmax(), 'title']

print(highest_points_to_price)

Bandit NV Merlot (California)


#### Question 6: ####
Create a series `flavor_counts` that contains two values: the number of wines that has the word "tart" in the `description` column and the number of wines that has the word "berries" in the `description` column. The index of the Series should be "Tart" and "Berries" for the corresponding values.

In [48]:
# Look for the word "tart", not case sensitive, not a substring of another word.
# Same applies to the word "berries"

flavor_counts = pd.Series({
    'Tart': wine['description'].str.contains('tart', case=False).sum(), ## tart -> case = False is searching while not case sensitive.
    'Berries': wine['description'].str.contains('berries', case=False).sum() ## berries 
})

flavor_counts.head()


Tart       4424
Berries    3512
dtype: int64

#### Question 7: ####
Let's convert the points into simple star ratings. A score of 90 or higher counts as 3 stars, a score of at least 80 but less than 90 is 2 stars. Any other score is 1 star.

Also, any wines from France should automatically get 3 stars, regardless of points.

Add this new column `star_ratings` to the dataframe with the number of stars for each wine in the dataset.

In [52]:
# 

def star_rating(row): # 
    if row['country'] == 'France': # wines from France automatically get 3 stars
        return 3
    elif row['points'] >= 90: # if scored higher than 90, 3 stars
        return 3
    elif row['points'] >= 80: # if scored at least 80 but less than 90, 2 stars
        return 2
    else:
        return 1 # otherwise, 1 star

wine['star_ratings'] = wine.apply(star_rating, axis=1) # create new column like before. Apply the star rating

wine.head(10) # use .head(10) to check more than 5 rows

country                                        description  \
0     Italy  Aromas include tropical fruit, broom, brimston...   
1  Portugal  This is ripe and fruity, a wine that is smooth...   
2        US  Tart and snappy, the flavors of lime flesh and...   
3        US  Pineapple rind, lemon pith and orange blossom ...   
4        US  Much like the regular bottling from 2012, this...   
5     Spain  Blackberry and raspberry aromas show a typical...   
6     Italy  Here's a bright, informal red that opens with ...   
7    France  This dry and restrained wine offers spice in p...   
8   Germany  Savory dried thyme notes accent sunnier flavor...   
9    France  This has great depth of flavor with its fresh ...   

                          designation  points  price           province  \
0                        Vulkà Bianco      87    NaN  Sicily & Sardinia   
1                            Avidagos      87   15.0              Douro   
2                                 NaN      87   14.0             Oregon   
3                Reserve Late Harvest      87   13.0           Michigan   
4  Vintner's Reserve Wild Child Block      87   65.0             Oregon   
5                        Ars In Vitro      87   15.0     Northern Spain   
6                             Belsito      87   16.0  Sicily & Sardinia   
7                                 NaN      87   24.0             Alsace   
8                               Shine      87   12.0        Rheinhessen   
9                         Les Natures      87   27.0             Alsace   

              region_1           region_2         taster_name  \
0                 Etna                NaN       Kerin O’Keefe   
1                  NaN                NaN          Roger Voss   
2    Willamette Valley  Willamette Valley        Paul Gregutt   
3  Lake Michigan Shore                NaN  Alexander Peartree   
4    Willamette Valley  Willamette Valley        Paul Gregutt   
5              Navarra                NaN   Michael Schachner   
6             Vittoria                NaN       Kerin O’Keefe   
7               Alsace                NaN          Roger Voss   
8                  NaN                NaN  Anna Lee C. Iijima   
9               Alsace                NaN          Roger Voss   

  taster_twitter_handle                                              title  \
0          @kerinokeefe                  Nicosia 2013 Vulkà Bianco  (Etna)   
1            @vossroger      Quinta dos Avidagos 2011 Avidagos Red (Douro)   
2           @paulgwine       Rainstorm 2013 Pinot Gris (Willamette Valley)   
3                   NaN  St. Julian 2013 Reserve Late Harvest Riesling ...   
4           @paulgwine   Sweet Cheeks 2012 Vintner's Reserve Wild Child...   
5           @wineschach  Tandem 2011 Ars In Vitro Tempranillo-Merlot (N...   
6          @kerinokeefe   Terre di Giurfo 2013 Belsito Frappato (Vittoria)   
7            @vossroger              Trimbach 2012 Gewurztraminer (Alsace)   
8                   NaN  Heinz Eifel 2013 Shine Gewürztraminer (Rheinhe...   
9            @vossroger  Jean-Baptiste Adam 2012 Les Natures Pinot Gris...   

              variety               winery     ratio  star_ratings  
0         White Blend              Nicosia       NaN             2  
1      Portuguese Red  Quinta dos Avidagos  5.800000             2  
2          Pinot Gris            Rainstorm  6.214286             2  
3            Riesling           St. Julian  6.692308             2  
4          Pinot Noir         Sweet Cheeks  1.338462             2  
5  Tempranillo-Merlot               Tandem  5.800000             2  
6            Frappato      Terre di Giurfo  5.437500             2  
7      Gewürztraminer             Trimbach  3.625000             3  
8      Gewürztraminer          Heinz Eifel  7.250000             2  
9          Pinot Gris   Jean-Baptiste Adam  3.222222             3

#### Question 8: ####
Who are the most common wine reviewers in the dataset? Create a Series whose index is the taster_twitter_handle category from the dataset, and whose values count how many reviews each person wrote.

In [63]:
# create taster_twitter_handle series 
taster_twitter_handle = wine['taster_twitter_handle'].value_counts() # .value_counts() creates a series already

#showing how many reviews each person wrote
print(taster_twitter_handle)

wine.head()

taster_twitter_handle
@vossroger          13045
@wineschach          7752
@kerinokeefe         5313
@paulgwine           4851
@vboone              4696
@mattkettmann        3035
@JoeCz               2605
@wawinereport        2358
@gordone_cellars     2032
@AnneInVino          1769
@laurbuzz             938
@suskostrzewa         593
@worldwineguys        465
@bkfiona               11
@winewchristina         4
Name: count, dtype: int64


country                                        description  \
0     Italy  Aromas include tropical fruit, broom, brimston...   
1  Portugal  This is ripe and fruity, a wine that is smooth...   
2        US  Tart and snappy, the flavors of lime flesh and...   
3        US  Pineapple rind, lemon pith and orange blossom ...   
4        US  Much like the regular bottling from 2012, this...   

                          designation  points  price           province  \
0                        Vulkà Bianco      87    NaN  Sicily & Sardinia   
1                            Avidagos      87   15.0              Douro   
2                                 NaN      87   14.0             Oregon   
3                Reserve Late Harvest      87   13.0           Michigan   
4  Vintner's Reserve Wild Child Block      87   65.0             Oregon   

              region_1           region_2         taster_name  \
0                 Etna                NaN       Kerin O’Keefe   
1                  NaN                NaN          Roger Voss   
2    Willamette Valley  Willamette Valley        Paul Gregutt   
3  Lake Michigan Shore                NaN  Alexander Peartree   
4    Willamette Valley  Willamette Valley        Paul Gregutt   

  taster_twitter_handle                                              title  \
0          @kerinokeefe                  Nicosia 2013 Vulkà Bianco  (Etna)   
1            @vossroger      Quinta dos Avidagos 2011 Avidagos Red (Douro)   
2           @paulgwine       Rainstorm 2013 Pinot Gris (Willamette Valley)   
3                   NaN  St. Julian 2013 Reserve Late Harvest Riesling ...   
4           @paulgwine   Sweet Cheeks 2012 Vintner's Reserve Wild Child...   

          variety               winery     ratio  star_ratings  
0     White Blend              Nicosia       NaN             2  
1  Portuguese Red  Quinta dos Avidagos  5.800000             2  
2      Pinot Gris            Rainstorm  6.214286             2  
3        Riesling           St. Julian  6.692308             2  
4      Pinot Noir         Sweet Cheeks  1.338462             2

The taster_twitter_handle shows that the most @vossroger did 13045 reviews, followed by @wineschach at 7752, @kerinokeefe at 5313, and @paulgwine at 4851.

#### Question 9: ####
What combination of countries and varieties are most common? Create a Series whose index is a MultiIndexof {country, variety} pairs. For example, a pinot noir produced in the US should map to {"US", "Pinot Noir"}. Sort the values in the Series in descending order based on wine count.

In [79]:
# MultiIndex is created with .groupby() with two tier pairs

country_variety_combo = wine.groupby(['country', 'variety']).size().sort_values(ascending=False) # ascending = False to make it descending order based on wine count

country_variety_combo.head(10)
country_variety_combo.to_frame()


0
country variety                       
US      Pinot Noir                4918
        Cabernet Sauvignon        3649
        Chardonnay                3412
France  Bordeaux-style Red Blend  2380
Italy   Red Blend                 1870
...                                ...
        Torbato                      1
        Vespaiolo                    1
        Vespolina                    1
        Vitovska                     1
Uruguay Tempranillo-Tannat           1

[1304 rows x 1 columns]

#### Question 10 #####
Create a Series whose index is reviewers and whose values is the average score given out by that reviewer.  
*Hint:* You will need the `taster_name` and `points` columns.

In [75]:
# use taster_name and points columns

reviewer_scores = wine.groupby('taster_name')['points'].mean()

reviewer_scores.to_frame()

points
taster_name                  
Alexander Peartree  86.014286
Anna Lee C. Iijima  88.380506
Anne Krebiehl MW    90.587903
Carrie Dykes        86.644444
Christina Pickard   89.500000
Fiona Adams         87.090909
Jeff Jenssen        88.273504
Jim Gordon          88.604331
Joe Czerwinski      88.519770
Kerin O’Keefe       88.827969
Lauren Buzzeo       87.831557
Matt Kettmann       90.021087
Michael Schachner   86.904541
Mike DeSimone       89.030303
Paul Gregutt        89.095032
Roger Voss          88.678957
Sean P. Sullivan    88.666243
Susan Kostrzewa     86.408094
Virginie Boone      89.229557